### PR TITLE
Refactor store index into its structure

### PR DESCRIPTION
**Pull request by wojtek-t**
```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

/cc @MadhavJivrajani

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go [hunk 1]
```diff
@@ -59,15 +59,89 @@ type ThreadSafeStore interface {
 	Resync() error
 }
 
+// storeIndex implements the indexing functionality for Store interface
+type storeIndex struct {
```
Comment (MadhavJivrajani) [inline staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go:63]: Nit: the definition and methods of this type are kind of split as of now (some defined here and some towards the end of the file) - if possible, it'd be nice to have them together.

**Review by MadhavJivrajani — COMMENTED** [review]
Looks great!
Just one non-blocking nit.

**Review by wojtek-t — COMMENTED** [review]
PTAL

**Review by MadhavJivrajani — APPROVED** [review]
/lgtm
/hold
Feel free to remove the hold when ready, thanks!

**Comment by pacoxu** [comment]
/retest
the unit failed for another flake

**Pull request merged**